File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int>visited(nums.size(),0);
        vector<vector<int>>ans;
        vector<int>temp;
        helper(nums,visited,temp,ans,0);
        return ans;
    }

    void helper(vector<int>& nums , vector<int> &visited,vector<int>&temp ,vector<vector<int>>&ans,int idx){

        if(visited.size() == temp.size()){
             ans.push_back(temp);
             return;
        }
        
        for(int i=0 ; i<visited.size(); i++){
             if(visited[i] == 0){
                visited[i] = 1;
                temp.push_back(nums[i]);
                helper(nums, visited, temp,ans,idx+1);
                visited[i] = 0;
                temp.pop_back();
             }
        }

    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    void swap(int a, int b, int[] nums){
      int temp = nums[a];
      nums[a] = nums[b];
      nums[b] = temp;
    }
    void help(int idx , int[] nums ,List<int[]> temp,List<List<Integer>> ans ){
     
     if(idx == nums.length){
       List<Integer> t =new ArrayList<>();

       for(int a: nums){
         t.add(a);
       }
       ans.add(new ArrayList<>(t));
        return;
     }

     for(int i=idx;i<nums.length;i++){
       swap(idx,i,nums);
      
       help(idx+1,nums,temp,ans);
       swap(idx,i,nums);
      
     }
     return;
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<int[]> temp = new ArrayList<>();
        help(0,nums,temp,ans);
        return ans;
    }
}